otter/convergence/steps.py

```diff
@@ -70,6 +70,46 @@ def set_server_name(server_config_args, name_suffix):
     return set_in(server_config_args, ('server', 'name'), name)
 
 
+def _failure_reporter(*terminal_err_types):
+    """
+    Return a callable that takes an error tuple which interprets the error
+    tuple.
+
+    If the error is an APIError with status code 4xx, or one of the provided
+    ``terminal_err_types``, then the callable returns a tuple of::
+
+        (StepResult.FAILURE, [ErrorReason.Exception(exc_tuple)])
+
+    else it returns a tuple of::
+
+        (StepResult.RETRY, [ErrorReason.Exception(exc_tuple)])
+    """
+    def reporter(exc_tuple):
+        err_type, error, traceback = exc_tuple
+
+        terminal_error = (
+            err_type in terminal_err_types or
+            (err_type == APIError and 400 <= error.code < 500)
+        )
+
+        if terminal_error:
+            return StepResult.FAILURE, [ErrorReason.Exception(exc_tuple)]
+        return StepResult.RETRY, [ErrorReason.Exception(exc_tuple)]
+
+    return reporter
+
+
+def _success_reporter(success_reason):
```

Comment:
I know this is because of a rather poor design we currently have in steps (and I take the blame for that), but it seems weird have something called `_success_reporter` be a thing that returns RETRY. I think it'd be preferable to call it `_retry_reporter`, though at some point in the future it'd be nice if we could get rid of the necessity of having most of the steps return RETRY.